File: leet_lint_code/lint_code/longest_increasing_subsequence.cpp
// Copyright © 2018 Hailin Gu. All rights reserved.
// License(GPL)
// Author: Hailin Gu
// This is a answer of lint code problem 76.

#ifndef LEETLINTCODE_LINTCODE_LONGESTINCREASINGSUBSEQUENCE_CPP
#define LEETLINTCODE_LINTCODE_LONGESTINCREASINGSUBSEQUENCE_CPP

#include <vector>
#include <map>

class LongestIncreasingSubsequence {
 public:
  int longestIncreasingSubsequence(std::vector<int> &nums) {
    int i = 0, len = (int)nums.size();
    if (0 == len) return 0;
    int tmp_cnt = 0;
    std::map<int, int> index_cnt;
    std::map<int, int>::iterator iter;
    while (i < len) {
      iter = index_cnt.begin();
      tmp_cnt = 1;
      while (iter != index_cnt.end()) {
        if (nums[i] > iter->first) {
          if (tmp_cnt < iter->second + 1)
            tmp_cnt = iter->second + 1;
        }
        ++ iter;
      }
      index_cnt[nums[i]] = tmp_cnt;
      ++ i;
    }
    
    int max = -1;
    for (std::map<int, int>::iterator iter = index_cnt.begin(); iter != index_cnt.end(); ++ iter)
      if (iter->second > max)
        max = iter->second;
    return max;
  }
};

#endif // LEETLINTCODE_LINTCODE_LONGESTINCREASINGSUBSEQUENCE_CPP

File: leet_lint_code/leet_code/java/LongestIncreasingSubsequence.java
// Copyright © 2018 Hailin Gu. All rights reserved.
// License
// Author: leetcode
// This is a answer of leet code problem 300.
// Date: 2019.8.21

class  LongestIncreasingSubsequence {
    public int lengthOfLIS(int[] nums) {
        if (nums.length == 0) {
            return 0;
        }
        
        if (nums.length == 1) {
            return 1;
        }
        
        int[] dp = new int[nums.length + 1];
        dp[0] = 1;
        int ans = Integer.MIN_VALUE;
        for (int i = 1; i < nums.length; ++ i) {
            dp[i] = 1;
            int j = i - 1;
            int dpt = 0;
            while(j >= 0) {
                if (nums[j] < nums[i]) {
                    dpt = Math.max(dpt, dp[j]);
                }
                
                -- j;
                
            }
            
            dp[i] = dpt + 1;
            ans = Math.max(dp[i], ans);
        }
        return ans;
    }
}